### PR TITLE
fixing small leak on exception on the transport-epoll-native allocation

### PR DESCRIPTION
**Pull request by clebertsuconic**
Motivation:

the JNI function ThrowNew won't release any allocated memory.
The method exceptionMessage is allocating a new string concatenating 2 constant strings
What is creating a small leak in case of these exceptions are happening.

Modifications:

Added new methods that will use exceptionMessage and free resources accordingly.
I am also removing the inline definition on these methods as they could be reused by
other added modules (e.g. libaio which should be coming soon)

Result:

No more leaks in case of failrues.

### CODE REVIEW

**Comment by netkins** [comment]
**AUTOMATED MESSAGE for ADMINS**:<br>Please verify and accept the pull request.<br>To accept, say `@netkins accept`<br>To build again, say `@netkins build`<br>To whitelist the author, say `@netkins whitelist`

**Comment by normanmaurer** [comment]
@netkins accept

File: transport-native-epoll/src/main/c/exception_helper.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2013 The Netty Project
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/exception_helper.h]: 2015

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c [hunk 2]
```diff
@@ -607,6 +636,9 @@ void JNI_OnUnload(JavaVM* vm, void* reserved) {
         if (netUtilClass != NULL) {
             (*env)->DeleteGlobalRef(env, netUtilClass);
         }
+
+        // unload hook from DirectFileDescriptorController
+        directFile_JNI_OnUnLoad(env);
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: remove changes which are not related to the fix please

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c [hunk 3]
```diff
@@ -578,6 +604,9 @@ jint JNI_OnLoad(JavaVM* vm, void* reserved) {
             return JNI_ERR;
         }
 
+        // load hook from DirectFileDescriptorController
+        directFile_JNI_OnLoad(env);
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: remove changes which are not related to the fix please

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c [hunk 4]
```diff
@@ -348,6 +370,10 @@ static int init_in_addr(JNIEnv* env, jbyteArray address, struct in_addr* addr) {
 }
 // util methods end
 
+// prototypes for the load hook from DirectFileDescriptorController.c
+void directFile_JNI_OnLoad(JNIEnv* env);
+void directFile_JNI_OnUnLoad(JNIEnv* env);
+
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: remove changes which are not related to the fix please

**Comment by normanmaurer** [comment]
@clebertsuconic can you please sign the ICLA:

http://tinyurl.com/netty-icla

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c (hunk 4 above)
Comment (clebertsuconic) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: geez... will do

**Comment by clebertsuconic** [comment]
I already did the cla.. about a month (3 weeks maybe) ago

**Comment by clebertsuconic** [comment]
@normanmaurer  I have made on my git copies to fix these diffs.. I probably created a thousand jobs on the netkins.. but I don't have authorization to cleanup them... 

I'm settled now.. if you could please cleanup the queue since I don't have authorization to kill a job.

**Comment by Scottmitch** [comment]
@clebertsuconic - Awesome! Thanks for the fix!

**Comment by normanmaurer** [comment]
@clebertsuconic thanks again!

Cherry-picked into 4.0 (ea0f84e72d5cab4f5793e6ee68b7f7e9ab5cde45), 4.1 (70e3d17620061767ce06f69e11a370fb61d69d08) and master (c0e889ae543e291f03f003485ee54dd281945ec6)
